packages/models/src/azure-openai/chat.ts

```diff
@@ -0,0 +1,281 @@
+import { POST, HttpError } from '@axflow/models/shared';
+import { streamTransformer } from '../openai/shared';
+
+export function headers(apiKey: string, customHeaders?: Record<string, string>) {
+  const headers: Record<string, string> = {
+    accept: 'application/json',
+    'content-type': 'application/json',
+    ...customHeaders,
+  };
+
+  if (typeof apiKey === 'string') {
+    headers['api-key'] = apiKey;
+  }
+
+  return headers;
+}
+
+// Our currently supported API version. Swagger spec:
+// https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-08-01-preview/inference.json
+const API_VERSION = '2023-08-01-preview';
+/*
+ * Create the base URL for the chat/completions endpoint in azure
+ * Docs: https://learn.microsoft.com/en-us/azure/ai-services/openai/reference
+ *
+ * We currently use the preview branch, and not the stable one, to benefit from openAI functions.
+ */
+const createUrl = (apiUrl: string | { resourceName: string; deploymentId: string }) => {
+  if (typeof apiUrl === 'string') {
+    return apiUrl;
+  } else {
+    return `https://${apiUrl.resourceName}.openai.azure.com/openai/deployments/${apiUrl.deploymentId}/chat/completions?api-version=${API_VERSION}`;
+  }
+};
+
+export namespace AzureOpenAIChatTypes {
+  export type RequestOptions = {
+    apiKey: string;
+    apiUrl: string | { resourceName: string; deploymentId: string };
```

Comment:
Can we accept a string and force users to construct it? That keeps it consistent with all the other modules, reduces number of options, and removes the conditional type / logic.

Reply:
I don't like the string much, because the types depend on the API version. So if they chose a string and pass the wrong version (pretty easy mistake), the types become wrong.

In addition, I personally found that the concepts for microsoft are a little counter-intuitive, so having the structured object with named values helps when you are searching for the inputs. When I asked Codestory for their url they gave me the wrong one, and I had to follow up asking for resource and deployment IDs.

I prefer to keep it as is

Reply:
Then let's remove the string rather than support both. Alternatively, we keep the string only, and we url parse and ensure the version is >= what we expect.